orttraining/orttraining/python/checkpointing_utils.py

```diff
@@ -1,143 +1,123 @@
 import os
-from collections import OrderedDict
 import torch
 
+
 def list_checkpoint_files(checkpoint_dir, checkpoint_prefix, extension='.ort.pt'):
     ckpt_file_names = [f for f in os.listdir(checkpoint_dir) if f.startswith(checkpoint_prefix)]
     ckpt_file_names = [f for f in ckpt_file_names if f.endswith(extension)]
     ckpt_file_names = [os.path.join(checkpoint_dir, f) for f in ckpt_file_names]
-    
+
     assert len(ckpt_file_names) > 0, "No checkpoint files found with prefix \"{}\" in directory {}.".format(checkpoint_prefix, checkpoint_dir)
     return ckpt_file_names
 
+
 def get_checkpoint_name(prefix, is_partitioned, world_rank=None, world_size=None):
-    SINGLE_CHECKPOINT_FILENAME='{prefix}.ort.pt'
-    MULTIPLE_CHECKPOINT_FILENAME='{prefix}.ZeRO.{world_rank}.{world_size}.ort.pt'
-    
+    SINGLE_CHECKPOINT_FILENAME = '{prefix}.ort.pt'
+    MULTIPLE_CHECKPOINT_FILENAME = '{prefix}.ZeRO.{world_rank}.{world_size}.ort.pt'
+
     if is_partitioned:
-        filename=MULTIPLE_CHECKPOINT_FILENAME.format(prefix=prefix, world_rank=world_rank, world_size=(world_size-1))
+        filename = MULTIPLE_CHECKPOINT_FILENAME.format(prefix=prefix, world_rank=world_rank, world_size=(world_size-1))
     else:
-        filename=SINGLE_CHECKPOINT_FILENAME.format(prefix=prefix)
-    
+        filename = SINGLE_CHECKPOINT_FILENAME.format(prefix=prefix)
+
     return filename
 
-class CombineZeroCheckpoint(object):
-    def __init__(self, checkpoint_files, clean_state_dict = None):
 
-        assert len(checkpoint_files) > 0, "No checkpoint files passed."
+def _split_state_dict(state_dict):
+    optimizer_keys = ['Moment_1_', 'Moment_2_', 'Update_Count_', 'Step_']
+    split_sd = {'optimizer': {}, 'fp32_param': {}, 'fp16_param': {}}
+    for k, v in state_dict.items():
+        mode = 'fp32_param'
+        for optim_key in optimizer_keys:
+            if k.startswith(optim_key):
+                mode = 'optimizer'
+                break
+        if k.endswith('_fp16'):
+            mode = 'fp16_param'
+        split_sd[mode][k] = v
+    return split_sd
```

Comment:
For the new design, our state dictionary has separate model and optimizer states. In addition, the fp32 and the fp16 states are also separate for both model and optimizer states. So, this entire function would no longer be needed since all the states are already split.
However, with the current design, we will not be able to completely remove the string matching that is happening in the function ```_split_name(self, name)```. This was the trade off we had to make when considering to save the relationship map and doing string manipulation. We will not be storing the complete relationship map (thereby saving on the save and load times), but will need to continue doing some amount of string matching and manipulation to aggregate the state dictionaries.

Reply:
I believe we still can do this without string matching. What is the reason BE cannot provide this information? Why is it not solvable?

Reply:
this should be addressed in the new design

Reply:
After a discussion with @ashbhandare , I believe we have a design that would work so that we do not do any string comparisons on the front end.